File: 5-mle/2-langgraph/1-langgraph-basic/graph-advanced/09_observability/metrics.py
"""토큰/비용 메트릭 헬퍼.

가격은 코드에 직접 박아두고, 실제 청구액과는 차이가 있을 수 있음을 README 에 명시.
"""
from __future__ import annotations

from typing import TypedDict


class MetricsRecord(TypedDict):
    model: str
    input_tokens: int
    output_tokens: int
    cost_usd: float


# Bedrock Anthropic 가격 (USD per 1,000 tokens). 정확한 청구액은 AWS 콘솔 기준.
# 출처: 작성 시점의 공개 가격표. 변경되면 이 값을 업데이트하면 끝.
PRICING: dict[str, dict[str, float]] = {
    "fast":  {"input": 0.001,  "output": 0.005},
    "default": {"input": 0.003,  "output": 0.015},
    "smart":   {"input": 0.015,  "output": 0.075},
}


def estimate_cost(model_alias: str, input_tokens: int, output_tokens: int) -> float:
    p = PRICING.get(model_alias, PRICING["default"])
    return (input_tokens / 1000.0) * p["input"] + (output_tokens / 1000.0) * p["output"]


def merge_metrics(a: list[MetricsRecord], b: list[MetricsRecord]) -> list[MetricsRecord]:
    """state reducer 가 필요할 때 사용 (현재 그래프는 명시적 append 사용)."""
    return list(a) + list(b)


__all__ = ["MetricsRecord", "PRICING", "estimate_cost", "merge_metrics"]

File: 5-mle/2-langgraph-js/apps/bff/src/langgraph-examples/graph-advanced/09-observability/metrics.ts
export interface ModelPricing {
  input: number;
  output: number;
}

export interface TokenUsage {
  inputTokens: number;
  outputTokens: number;
  totalTokens: number;
}

export interface MetricsRecord extends TokenUsage {
  model: string;
  costUsd: number;
}

export interface MetricsAggregate {
  metrics: MetricsRecord[];
  totalCostUsd: number;
  totalInputTokens: number;
  totalOutputTokens: number;
  totalTokens: number;
}

// Demo estimates in USD per 1,000 tokens. Update this table when provider prices change.
export const PRICING: Record<string, ModelPricing> = {
  fast: { input: 0.001, output: 0.005 },
  default: { input: 0.003, output: 0.015 },
  normal: { input: 0.003, output: 0.015 },
  smart: { input: 0.015, output: 0.075 },
  reasoning: { input: 0.015, output: 0.075 },
  "gpt-4o-mini": { input: 0.001, output: 0.005 }
};

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function readNumber(record: Record<string, unknown>, keys: readonly string[]): number | undefined {
  for (const key of keys) {
    const value = record[key];
    if (typeof value === "number") {
      return value;
    }
    if (typeof value === "string" && value.trim() !== "") {
      const parsed = Number(value);
      if (Number.isFinite(parsed)) {
        return parsed;
      }
    }
  }
  return undefined;
}

function normalizeTokenCount(value: number | undefined, label: string): number {
  if (value === undefined) {
    return 0;
  }
  if (!Number.isFinite(value) || value < 0) {
    throw new RangeError(`${label} must be a non-negative finite number.`);
  }
  return Math.trunc(value);
}

function readUsageFrom(record: Record<string, unknown>): TokenUsage {
  const inputTokens = normalizeTokenCount(
    readNumber(record, ["input_tokens", "inputTokens", "prompt_tokens", "promptTokens"]),
    "inputTokens"
  );
  const outputTokens = normalizeTokenCount(
    readNumber(record, ["output_tokens", "outputTokens", "completion_tokens", "completionTokens"]),
    "outputTokens"
  );
  const explicitTotal = readNumber(record, ["total_tokens", "totalTokens"]);
  const totalTokens = normalizeTokenCount(explicitTotal ?? inputTokens + outputTokens, "totalTokens");

  return {
    inputTokens,
    outputTokens,
    totalTokens
  };
}

export function estimateCost(modelAlias: string, inputTokens: number, outputTokens: number): number {
  const input = normalizeTokenCount(inputTokens, "inputTokens");
  const output = normalizeTokenCount(outputTokens, "outputTokens");
  const pricing = PRICING[modelAlias] ?? PRICING.default;
  return Number(((input / 1000) * pricing.input + (output / 1000) * pricing.output).toFixed(12));
}

export function extractUsage(source: unknown): TokenUsage {
  if (!isRecord(source)) {
    return {
      inputTokens: 0,
      outputTokens: 0,
      totalTokens: 0
    };
  }

  if (isRecord(source.usage_metadata)) {
    return readUsageFrom(source.usage_metadata);
  }

  if (isRecord(source.response_metadata)) {
    const tokenUsage = source.response_metadata.tokenUsage ?? source.response_metadata.token_usage;
    if (isRecord(tokenUsage)) {
      return readUsageFrom(tokenUsage);
    }
  }

  return readUsageFrom(source);
}

export function createMetricsRecord(modelAlias: string, source: unknown): MetricsRecord {
  const usage = extractUsage(source);
  return {
    model: modelAlias,
    ...usage,
    costUsd: estimateCost(modelAlias, usage.inputTokens, usage.outputTokens)
  };
}

export function mergeMetrics(left: readonly MetricsRecord[], right: readonly MetricsRecord[]): MetricsRecord[] {
  return [...left, ...right];
}

export function summarizeMetrics(metrics: readonly MetricsRecord[]): MetricsAggregate {
  const totalInputTokens = metrics.reduce((sum, metric) => sum + metric.inputTokens, 0);
  const totalOutputTokens = metrics.reduce((sum, metric) => sum + metric.outputTokens, 0);
  const totalTokens = metrics.reduce((sum, metric) => sum + metric.totalTokens, 0);
  const totalCostUsd = Number(metrics.reduce((sum, metric) => sum + metric.costUsd, 0).toFixed(12));

  return {
    metrics: [...metrics],
    totalCostUsd,
    totalInputTokens,
    totalOutputTokens,
    totalTokens
  };
}

export function appendMetric(metrics: readonly MetricsRecord[], record: MetricsRecord): MetricsAggregate {
  return summarizeMetrics([...metrics, record]);
}
